addon/lib/webextension-channels.js

```diff
@@ -0,0 +1,149 @@
+/*
+ * This Source Code is subject to the terms of the Mozilla Public License
+ * version 2.0 (the 'License'). You can obtain a copy of the License at
+ * http://mozilla.org/MPL/2.0/.
+ */
+
+/* global XPCOMUtils, Services */
+
+const { Ci, Cu } = require('chrome');
+const { Class } = require('sdk/core/heritage');
+const { Disposable } = require('sdk/core/disposable');
+
+const store = require('sdk/simple-storage').storage;
+
+const Metrics = require('./metrics');
+
+const TESTPILOT_TELEMETRY_CHANNEL = 'testpilot-telemetry';
+
+Cu.import('resource://gre/modules/XPCOMUtils.jsm');
+
+XPCOMUtils.defineLazyModuleGetter(this, 'Services',
+                                  'resource://gre/modules/Services.jsm');
+
+const {getExtensionUUID} = Cu.import('resource://gre/modules/Extension.jsm', {});
+
+function createChannelForAddonId(name, addonId) {
+  const targetExtensionUUID = getExtensionUUID(addonId);
+  const baseURI = Services.io
+        .newURI(`moz-extension://${targetExtensionUUID}/_blank.html`, null, null);
+  const principal = Services.scriptSecurityManager
+        .createCodebasePrincipal(baseURI, { addonId });
+
+  const addonChromeWebNav = Services.appShell.createWindowlessBrowser(true);
+  const docShell = addonChromeWebNav.QueryInterface(Ci.nsIInterfaceRequestor) // eslint-disable-line new-cap
+        .getInterface(Ci.nsIDocShell);
+  docShell.createAboutBlankContentViewer(principal);
+  const window = docShell.contentViewer.DOMDocument.defaultView;
+  const addonBroadcastChannel = new window.BroadcastChannel(name);
+
+  return {
+    addonChromeWebNav,
+    addonBroadcastChannel
+  };
+}
+
+const WebExtensionChannel = Class({ // eslint-disable-line new-cap
+  implements: [Disposable],
+
+  initialize(targetAddonId) {
+    this.pingListeners = new Set();
+
+    this.targetAddonId = targetAddonId;
+
+    const {
+      addonChromeWebNav,
+      addonBroadcastChannel
+    } = createChannelForAddonId(TESTPILOT_TELEMETRY_CHANNEL, targetAddonId);
+
+    // NOTE: Keep a ref to prevent it from going away during garbage collection
+    // (or the BroadcastChannel will stop working).
+    this.addonChromeWebNav = addonChromeWebNav;
+    this.addonBroadcastChannel = addonBroadcastChannel;
+
+    this.handleEventBound = ev => this.handleEvent(ev);
```

Comment:
I could probably just do `this.handleEvent.bind(this)`, but thought this looked clearer. I think the add/remove event methods need a consistent bound function ref to deal with. Though I *did* think `handleEvent` and passing `this` was enough.